pkg/license/license.go

```diff
@@ -0,0 +1,329 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package license
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"chainguard.dev/melange/pkg/config"
+	"github.com/chainguard-dev/clog"
+	golicenses "github.com/google/go-licenses/v2/licenses"
+	licenseclassifier "github.com/google/licenseclassifier/v2"
+	"github.com/google/licenseclassifier/v2/assets"
+)
+
+// Classifier can detect the type of a software license.
+type Classifier interface {
+	Identify(fsys fs.FS, licensePath string) ([]License, error)
+}
+
+type melangeClassifier struct {
+	classifier *licenseclassifier.Classifier
+}
+
+// NewClassifier creates a license classifier.
+func NewClassifier() (Classifier, error) {
+	c, err := assets.DefaultClassifier()
+	if err != nil {
+		return nil, err
+	}
+	return &melangeClassifier{classifier: c}, nil
+}
+
+// License represents a software license, as detected by licenseclassifier.
+type License struct {
+	Name       string
+	Type       golicenses.Type
+	Confidence float64
+	Source     string
+}
+
+// LicenseFile represents a license file, with its name, path, and relevance score.
+type LicenseFile struct {
+	Name   string
+	Path   string
+	Weight float64
+}
+
+// LicenseDiff represents a difference between the detected license and the expected license.
+type LicenseDiff struct {
+	Path   string
+	Is     string
+	Should string
+}
+
+// Identify identifies the license of a file on a filesystem using the licenseclassifier.
+func (c *melangeClassifier) Identify(fsys fs.FS, licensePath string) ([]License, error) {
+	if licensePath == "" {
+		return nil, nil
+	}
+
+	file, err := fsys.Open(licensePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	matches, err := c.classifier.MatchFrom(file)
+	if err != nil {
+		return nil, err
+	}
+
+	foundLicenseNames := map[string]struct{}{}
+
+	licenses := []License{}
+	for _, match := range matches.Matches {
+		if match.MatchType != "License" {
+			continue
+		}
+
+		// Skip duplicate licenses.
+		if _, ok := foundLicenseNames[match.Name]; ok {
+			continue
+		}
+		foundLicenseNames[match.Name] = struct{}{}
+
+		licenses = append(licenses, License{
+			Name:       match.Name,
+			Type:       golicenses.LicenseType(match.Name),
+			Confidence: match.Confidence,
+			Source:     licensePath,
+		})
+	}
+
+	return licenses, nil
+}
+
+// FindLicenseFiles returns a list of license files in a directory, sorted by their relevance score.
+func FindLicenseFiles(fsys fs.FS) ([]LicenseFile, error) {
+	// Copied and adjusted from Ruby to Go from the "licensee" project
```

Comment:
do we need to include the MIT license text?

Reply:
if so, might make sense to move this function into a separate file to make the attribution clear